### ISSUE TITLE
any way to include non-compiled assets for inject manifest plugin?

### ISSUE THREAD

**Issue by novaknole** [GoogleChrome/workbox#2361]
**Library Affected**:
*workbox-webpack-plugin.*

**Browser & Platform**:
Latest

**Issue or Feature Request Description**:
I have 2 javascript files. Seems like I don't want webpack to compile them so I never include them in any js file. So the way I am doing it is I wrote my own plugin which has the very early hook such as thisCompile hook which does some things and then moves some files to /dist folder. Then, I know that `injectManifest` plugin uses `make/emit` hooks and gathers all things from there that pass the validation and creates manifest depending on it. The thing is the way workbox-webpack-plugin does that is it uses `compilation` object. Since my files that I moved in `thisCompile` hook of my plugin are not included in compilation assets, inject manifest won't know about these files and i am left with no manifest for these 2 files. 

I got some ideas, but maybe you can shoot the best practice about `injectManifest` plugin. I am using its latest version.

**Comment by jeffposnick** [comment]
The recommended approach for most folks would be to use https://github.com/webpack-contrib/copy-webpack-plugin to handle copying the existing assets over to the output directory. That should be enough to get the assets into `webpack`'s build pipeline, and `workbox-webpack-plugin` can pick them up from there.

If you can't do that, for some reason, you can add them in to the final manifest in Workbox v5 by using a configuration like

```
{
  additionalManifestEntries: [{
    url: '/path/to/my/asset.js', revision: 'YOUR_HASH_HERE',
  }. {
    url: '/path/to/my/versioned-asset.1234abcd.js', revision: null,
  }],
  // ...other Workbox configuration options...
}
```

If your assets don't contain hashes in their URLs (the first example), it's up to you to make sure you generate a unique hash of the file's contents and add it in as the `revision` parameter. If your assets do already contain hashes, you can safely set `revision: null`, as in the second example.

**Comment by novaknole** [comment]
The bad thing is that before copying, i have to do some transformation for the files which require some loops and custom logic. That's why I am using my custom plugin with `thisCompile` hook. but this way, workbox-webpack-plugin won't take these files as they won't exist in compilation's assets, but still will be in /dist folder. makes sense?

**Comment by jeffposnick** [comment]
Sure. Given that, you can use the second option, with `additionalManifestEntries`. If you need more flexibility, another option is to use the `manifestTransforms` option, which lets you specify one or more [functions](manifestTransforms) that can manipulation the precache manifest's contents. Inside that function, you can read from the filesystem, refer to the output created by your other plugin, etc.

**Comment by novaknole** [comment]
Awesome. could you quickly write how i can use it? couldn't find it in the docs.

```javascript

 new InjectManifest({
            swSrc: './src/js/service-worker.js',
            swDest: './service-worker.js',
            exclude: [
                /\.map$/,
                /^manifest.*\.js$/,
                /^.htaccess/
            ],
            manifestTransforms: [
                test(entries, compi) {
                    console.log(manif, ' ah')
                }
            ]
        })
```
this results in an error.

**Comment by novaknole** [comment]
@jeffposnick , do you recommend transforming manifest so that we remove revision for those urls which include hashes ? if yes/no, why?

Thanks.

**Comment by jeffposnick** [comment]
In general, using `revision: null` when a `url` already contains revision info (i.e. a hash) will lead less data being transferred via the network during the preaching process—assuming that you also are setting long-lived `Cache-Control` headers on your server for those URLs.

(The easiest way to ensure that `revision: null` is set for those URLs is to use the `dontCacheBustURLsMatching` parameter when [configuring your Workbox plugin](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-webpack-plugin.InjectManifest#InjectManifest). This will only work for manifest entries that are derived from the `webpack` asset list, which obviously won't apply to the assets that only your custom plugin knows about.)

**Comment by novaknole** [comment]
@jeffposnick  Thanks for the info, Appreciate it.

Meanwhile, I saw your answer on stackoverflow which states why skipWaiting is risky. https://stackoverflow.com/questions/51715127/what-are-the-downsides-to-using-skipwaiting-and-clientsclaim-with-workbox

I think the first scenario you're describing is not that much of a huge risk. This risk only happens if we have more than 1 tab open. And when an update to a service worker is uploaded, we should refresh one tab to install/activate it. Now, the problem is in the first tab,  if you click the link, view-one.hashbefore.js won't be there in a cache as second tab's refresh removed it. Now, what will happen is view-one.hashbefore.js 's request will get made to the network, which will return 404 as it's removed from the server. 

So to sum up, skipWaiting is a problem when one tab refreshes while new service worker update is on the server, and there's another tab which we click on links or whatever and the first tab's results happen to be broken as they don't exist on the server?

Agree?

**Comment by eolamisangoatote** [comment]
Here's an article how to scan the output folder and add anything you see fit into the `additionalManifestEntries` attribute.
Relevant code sample:

```
const fs = require('fs');
const globby = require('globby');
const crypto = require('crypto');
...
const getRevision = file => crypto.createHash('sha256').update(fs.readFileSync(file)).digest('hex');
...
const additionalManifestEntries = globby.sync([
  path.posix.join(outputDirPath, '/your_folder1'),
  ...
  path.posix.join(outputDirPath, '/your_folderX')]).map(f => ({
    url: f.replace(path.posix.join(outputDirPath, '/'), ''),
    revision: getRevision(f)
  }));
...
  new WorkboxPlugin.InjectManifest({
        additionalManifestEntries: additionalManifestEntries,
...
```

Source:
[Precaching pages with next-pwa - DEV Community](https://dev.to/sfiquet/precaching-pages-with-next-pwa-31f2)
